### PR TITLE
fm-list-model: Use a modern GObject definition for FMListModel

### PR DESCRIPTION
**Pull request by tintou**
Also add several sanity checks to all functions and methods.

### CODE REVIEW

File: libcore/fm-list-model.c [hunk 1]
```diff
@@ -839,25 +860,34 @@ fm_list_model_file_changed (FMListModel *model, GOFFile *file,
 gboolean
 fm_list_model_is_empty (FMListModel *model)
 {
-    return (g_sequence_get_length (model->details->files) == 0);
+    FMListModelPrivate *priv = fm_list_model_get_instance_private (model);
+
+    g_return_val_if_fail (FM_IS_LIST_MODEL (model), TRUE);
```
Comment (jeremypw) [inline libcore/fm-list-model.c:865]: Shouldn't this come first?

File: libcore/fm-list-model.c [hunk 2]
```diff
@@ -839,25 +860,34 @@ fm_list_model_file_changed (FMListModel *model, GOFFile *file,
...
+    return (g_sequence_get_length (priv->files) == 0);
 }
 
 guint
 fm_list_model_get_length (FMListModel *model)
 {
-    return g_sequence_get_length (model->details->files);
+    FMListModelPrivate *priv = fm_list_model_get_instance_private (model);
+
+    g_return_val_if_fail (FM_IS_LIST_MODEL (model), 0);
```
Comment (jeremypw) [inline libcore/fm-list-model.c:875]: Should come first?

File: libcore/fm-list-model.c [hunk 3]
```diff
@@ -946,45 +981,49 @@ fm_list_model_clear_directory (FMListModel *model, GSequence *files)
...
     }
 }
 
 void
 fm_list_model_clear (FMListModel *model)
 {
-    g_return_if_fail (model != NULL);
+    FMListModelPrivate *priv = fm_list_model_get_instance_private (model);
+
+    g_return_if_fail (FM_IS_LIST_MODEL (model));
```
Comment (jeremypw) [inline libcore/fm-list-model.c:994]: Should come first?

File: libcore/fm-list-model.c [hunk 4]
```diff
@@ -1226,17 +1258,21 @@ fm_list_model_sortable_init (GtkTreeSortableIface *iface)
...
     g_return_if_fail (FM_IS_LIST_MODEL (model));
 
-    model->details->has_child = has_child;
+    priv->has_child = has_child;
 }
 
 static void
 fm_list_model_set_icon_size (FMListModel *model, gint size)
 {
+    FMListModelPrivate *priv = fm_list_model_get_instance_private (model);
```
Comment (jeremypw) [inline libcore/fm-list-model.c:1271]: Should come after testing model is FMListModel?

**Comment by tintou** [comment]
@jeremypw foo_bar_get_instance_private is just a macro doing pointer offsets, it is safe to do it before testing the object.

**Comment by jeremypw** [comment]
This does not seem to leak any more memory than master when navigating around, creating and destroying models and no regressions found so far so looking good. I'll test a while before approving though.

**Review by jeremypw — APPROVED** [review]
AFAICT as a C non-expert the new code is equivalent to the old so not regressions are expected.  No regressions were found on limited testing - I think this is safe to merge.

**Pull request merged**